File: controladores/juegos/rosco.py
from datetime import datetime

from db.firebase import db
from modelos.juegos.rosco import SesionRosco
from servicios.rosco_servicio import (
    generar_preguntas_demo,
    obtener_siguiente_letra,
    calcular_puntaje
)


def iniciar_partida(id_usuario: str, tematica: str) -> dict:

    preguntas = generar_preguntas_demo()

    id_sesion = f"rosco_{id_usuario}_{int(datetime.utcnow().timestamp())}"

    sesion = SesionRosco(
        id_usuario=id_usuario,
        tematica=tematica,
        preguntas=preguntas
    )

    db.collection(
        "partidas_rosco"
    ).document(
        id_sesion
    ).set(
        sesion.to_dict()
    )

    return {
        "id_sesion": id_sesion,
        "tematica": tematica,
        "letra": "A",
        "definicion": preguntas["A"]["definicion"]
    }


def obtener_sesion(id_sesion: str) -> dict:

    doc = db.collection(
        "partidas_rosco"
    ).document(
        id_sesion
    ).get()

    if not doc.exists:
        return {"error": "Sesión no encontrada"}

    sesion = SesionRosco.from_dict(
        doc.to_dict()
    )

    data = {
        "id_sesion": id_sesion,
        "tematica": sesion.tematica,
        "estado": sesion.estado,
        "puntaje": sesion.puntaje,
        "aciertos": sesion.aciertos,
        "errores": sesion.errores,
        "letra_actual": sesion.letra_actual,
        "preguntas": sesion.preguntas
    }

    if sesion.letra_actual:
        data["definicion"] = (
            sesion.preguntas[
                sesion.letra_actual
            ]["definicion"]
        )

    return data


def responder_pregunta(
    id_sesion: str,
    letra: str,
    respuesta: str
) -> dict:

    doc = db.collection(
        "partidas_rosco"
    ).document(
        id_sesion
    ).get()

    if not doc.exists:
        return {"error": "Sesión no encontrada"}

    sesion = SesionRosco.from_dict(
        doc.to_dict()
    )

    if sesion.estado != "jugando":
        return {"error": "La partida ya terminó"}

    pregunta = sesion.preguntas[letra]

    respuesta = respuesta.upper().strip()

    correcta = (
        respuesta ==
        pregunta["respuesta"]
    )

    respuesta_correcta = None

    if correcta:
        pregunta["estado"] = "correcta"
        sesion.aciertos += 1
    else:
        pregunta["estado"] = "incorrecta"
        sesion.errores += 1
        respuesta_correcta = pregunta["respuesta"]

    siguiente_letra = obtener_siguiente_letra(
        sesion.preguntas
    )

    sesion.letra_actual = siguiente_letra

    if siguiente_letra is None:
        sesion.estado = "finalizada"
        sesion.puntaje = calcular_puntaje(
            sesion.aciertos
        )
        sesion.finalizada_en = (
            datetime.utcnow().isoformat()
        )

    db.collection(
        "partidas_rosco"
    ).document(
        id_sesion
    ).set(
        sesion.to_dict()
    )

    resultado = {
        "correcta": correcta,
        "respuesta_correcta": respuesta_correcta,
        "estado": sesion.estado,
        "aciertos": sesion.aciertos,
        "errores": sesion.errores,
        "puntaje": sesion.puntaje,
        "preguntas": sesion.preguntas
    }

    if siguiente_letra:
        resultado["letra_actual"] = siguiente_letra
        resultado["definicion"] = (
            sesion.preguntas[
                siguiente_letra
            ]["definicion"]
        )

    return resultado


def pasar_pregunta(
    id_sesion: str,
    letra: str
) -> dict:

    doc = db.collection(
        "partidas_rosco"
    ).document(
        id_sesion
    ).get()

    if not doc.exists:
        return {"error": "Sesión no encontrada"}

    sesion = SesionRosco.from_dict(
        doc.to_dict()
    )

    if sesion.estado != "jugando":
        return {"error": "La partida ya terminó"}

    sesion.preguntas[letra]["estado"] = "pasada"

    siguiente_letra = obtener_siguiente_letra(
        sesion.preguntas
    )

    sesion.letra_actual = siguiente_letra

    if siguiente_letra is None:
        sesion.estado = "finalizada"
        sesion.puntaje = calcular_puntaje(
            sesion.aciertos
        )
        sesion.finalizada_en = (
            datetime.utcnow().isoformat()
        )

    db.collection(
        "partidas_rosco"
    ).document(
        id_sesion
    ).set(
        sesion.to_dict()
    )

    resultado = {
        "estado": sesion.estado,
        "aciertos": sesion.aciertos,
        "errores": sesion.errores,
        "puntaje": sesion.puntaje,
        "preguntas": sesion.preguntas
    }

    if siguiente_letra:
        resultado["letra_actual"] = siguiente_letra
        resultado["definicion"] = (
            sesion.preguntas[
                siguiente_letra
            ]["definicion"]
        )

    return resultado

File: static/js/rosco.js
document.addEventListener("DOMContentLoaded", () => {

    inicializarRosco();

    window.addEventListener("resize", renderizarRosco);

    document
        .getElementById("btn-responder")
        .addEventListener(
            "click",
            responderPregunta
        );

    document
        .getElementById("respuesta-input")
        .addEventListener(
            "input",
            () => {
                document.getElementById(
                    "error-juego"
                ).textContent = "";
            }
        );

    document
        .getElementById("btn-pasar")
        .addEventListener(
            "click",
            pasarPregunta
        );

    document
        .getElementById("respuesta-input")
        .addEventListener(
            "keydown",
            e => {
                if (e.key === "Enter") {
                    responderPregunta();
                }
            }
        );

    document
        .getElementById("btn-volver-inicio")
        .addEventListener(
            "click",
            () => {
                window.location.href = "/";
            }
        );
});

function inicializarRosco() {
    document.getElementById(
        "tematica-actual"
    ).textContent = SESION.tematica;

    actualizarPantalla();
    renderizarRosco();
}

function actualizarPantalla() {
    document.getElementById(
        "letra-actual"
    ).textContent = SESION.letra_actual || "-";

    document.getElementById(
        "pregunta-actual"
    ).textContent = SESION.definicion || "";

    document.getElementById(
        "correctas"
    ).textContent = SESION.aciertos || 0;

    document.getElementById(
        "incorrectas"
    ).textContent = SESION.errores || 0;

    document.getElementById(
        "score-valor"
    ).textContent = SESION.puntaje || 0;
}

function renderizarRosco() {
    const contenedor = document.getElementById("rosco");
    if (!contenedor) return;

    contenedor.innerHTML = "";

    const letras = Object.keys(SESION.preguntas);
    if (letras.length === 0) return;

    const anchoContenedor = contenedor.clientWidth || 650; 
    
    const centro = anchoContenedor / 2;
    
    const radio = anchoContenedor * 0.40;

    letras.forEach((letra, index) => {
        const pregunta = SESION.preguntas[letra];
        const nodo = document.createElement("div");

        nodo.classList.add("letra");
        contenedor.appendChild(nodo); 

        const mitadLetra = nodo.offsetWidth / 2 || 25; 

        const angulo =
            ((Math.PI * 2) / letras.length) * index - Math.PI / 2;

        const x = centro + radio * Math.cos(angulo) - mitadLetra;
        const y = centro + radio * Math.sin(angulo) - mitadLetra;

        nodo.style.left = `${x}px`;
        nodo.style.top = `${y}px`;
        nodo.textContent = letra;

        if (letra === SESION.letra_actual) {
            nodo.classList.add("letra-activa");
        }

        if (pregunta.estado === "correcta") {
            nodo.classList.add("letra-correcta");
        }

        if (pregunta.estado === "incorrecta") {
            nodo.classList.add("letra-incorrecta");
        }

        if (pregunta.estado === "pasada") {
            nodo.classList.add("letra-pasada");
        }
    });
}

async function responderPregunta() {
    const respuesta = document.getElementById("respuesta-input").value.trim();

    if (!respuesta) {
        mostrarError("Ingresá una respuesta");
        setTimeout(() => {
            document.getElementById("error-juego").textContent = "";
        }, 3000);
        return;
    }

    const res = await fetch("/juegos/rosco/responder", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            id_sesion: SESION.id_sesion,
            letra: SESION.letra_actual,
            respuesta
        })
    });

    const data = await res.json();

    if (data.error) {
        mostrarError(data.error);
        return;
    }

    actualizarSesion(data);
    document.getElementById("respuesta-input").value = "";
}

async function pasarPregunta() {
    const res = await fetch("/juegos/rosco/pasar", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            id_sesion: SESION.id_sesion,
            letra: SESION.letra_actual
        })
    });

    const data = await res.json();

    if (data.error) {
        mostrarError(data.error);
        return;
    }

    actualizarSesion(data);
}

function actualizarSesion(data) {
    SESION.preguntas = data.preguntas;
    SESION.estado = data.estado;
    SESION.puntaje = data.puntaje;
    SESION.aciertos = data.aciertos;
    SESION.errores = data.errores;
    SESION.letra_actual = data.letra_actual;
    SESION.definicion = data.definicion;

    const respuestaCorrecta = document.getElementById("respuesta-correcta");

    if (data.correcta === false && data.respuesta_correcta) {
        respuestaCorrecta.textContent = `La respuesta correcta era: ${data.respuesta_correcta}`;
    } else {
        respuestaCorrecta.textContent = "";
    }

    actualizarPantalla();
    renderizarRosco();

    if (data.estado === "finalizada") {
        mostrarResultado(data);
    }
}

function mostrarResultado(data) {
    document.querySelector(".rosco-container").style.display = "none";
    const pantalla = document.getElementById("pantalla-resultado");
    pantalla.style.display = "flex";

    document.getElementById("resultado-titulo").textContent = "Rosco finalizado";
    document.getElementById("resultado-puntaje").textContent = `Puntaje: ${data.puntaje}`;
    document.getElementById("resultado-correctas").textContent = `Correctas: ${data.aciertos}`;
    document.getElementById("resultado-incorrectas").textContent = `Incorrectas: ${data.errores}`;
}

function mostrarError(msg) {
    const el = document.getElementById("error-juego");
    el.textContent = msg;
    el.classList.add("visible");
}